Lightweight solution load: Merge feature branch

alpaix opened the pull request:
This PR covers the feature work to enable LSL support as proposed in the spec document [Full support of LSL](https://github.com/NuGet/Home/wiki/Full-support-of-LSL).

Main concepts introduced

- `IVsProjectAdapter` - An abstraction layer over `EnvDTE.Project`.
Represents a virtual project object. It is aware of LSL mode and is
capable to force load project when it's needed. All occurances of DTE
project have been replaced with new adapter object.

- `INuGetProjectServices` - Collection of project services as building
blocks to access underlying project system functionality, such as
reference reader, build properties, project system capabilites, script host.

- `VsCoreProjectSystemService` - Implementation of project services for VS
  core project system (DTE).

- `VsLangProjectSystemService` - Implementation of project services for
  managed languages project system (VsLangProj80).

- `WorkspaceProjectServices` - Implementation of project services for
  deferred projects based on Workspace* API.

Notable refactorings
- Decommissioned `EnvDTEProjectAdapter` in favour of the new adapter.
- `LegacyCSProjPackageReferenceProject` =>
  `LegacyPackageReferenceProject`.
- `CpsPackageReferenceProject` => `NetCorePackageReferenceProject`
- Multiple renamings to avoid confusion betwen projects and project
  systems.

Resolves NuGet/Home#5180.
Resolves NuGet/Home#3711.

jainaashish commented:
Please add a short summary about the code implementation to give some background. Or a short spec will also be fine.

alpaix commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VsProject/DefaultProjectThreadingService.cs:
```diff
@@ -0,0 +1,34 @@
+            return NuGetUIThreadHelper.JoinableTaskFactory.Run(asyncMethod);
+        }
+
+        public void VerifyOnUIThread(string callerMemberName)
```
>VerifyOnUIThread [](start = 20, length = 16)

Keep the original method name

alpaix commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/LegacyPackageReferenceProject.cs:
```diff
@@ -0,0 +1,298 @@
+    /// </summary>
+    public sealed class LegacyPackageReferenceProject : BuildIntegratedNuGetProject
+    {
+        private readonly IVsProjectAdapter _vsProjectAdapter;
```
>IVsProjectAdapter [](start = 25, length = 17)

Consider moving to a base class

alpaix committed:
Lightweight solution load: Merge feature branch

This change summarizes feature work done to enable full support of LSL
in NuGet Package Manager.

Main concepts introduced

- `IVsProjectAdapter` - An abstraction layer over `EnvDTE.Project`.
Represents a virtual project object. It is aware of LSL mode and is
capable to force load project when it's needed. All occurances of DTE
project have been replaced with new adapter object.

- `INuGetProjectServices` - Collection of project services as building
blocks to access underlying project system functionality, such as
reference reader, build properties, project system capabilites, script host.

- `VsCoreProjectSystemService` - Implementation of project services for VS
  core project system (DTE).

- `VsLangProjectSystemService` - Implementation of project services for
  managed languages project system (VsLangProj80).

- `WorkspaceProjectServices` - Implementation of project services for
  deferred projects based on Workspace* API.

Notable refactorings
- Decommissioned `EnvDTEProjectAdapter` in favour of the new adapter.
- `LegacyCSProjPackageReferenceProject` =>
  `LegacyPackageReferenceProject`.
- `CpsPackageReferenceProject` => `NetCorePackageReferenceProject`
- Multiple renamings to avoid confusion betwen projects and project
  systems.

Resolves NuGet/Home#5180.

jainaashish commented on src/NuGet.Clients/NuGet.Console/ScriptExecutor.cs:
```diff
@@ -22,26 +22,31 @@
+        private AsyncLazy<IHost> Host { get; }
+
+        [Import]
+        private ISettings Settings { get; set; }
```
>ISettings [](start = 16, length = 9)

lazy?

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.PowerShellCmdlets/NuGetPowerShellBaseCommand.cs:
```diff
@@ -474,26 +474,23 @@ protected void GetNuGetProject(string projectName = null)
         /// Get default project in the type of EnvDTE.Project, to keep PowerShell scripts backward-compatbility.
         /// </summary>
```
Update comment to reflect the change

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.PowerShellCmdlets/Cmdlets/GetProjectCommand.cs:
```diff
@@ -50,24 +52,26 @@ protected override void ProcessRecordCore()
             {
-                var projects = EnvDTESolutionUtility.GetAllEnvDTEProjects(DTE);
+                VsSolutionManager.EnsureSolutionIsLoaded();
+                var projects = VsSolutionManager.GetAllVsProjectAdapters().Select(p => p.Project);
```
why we need this here? wouldn't it then force it to load all projects ?

alpaix replied:
With `-all` switch it retrieves all of DTE projects. SO it's more efficient to force entire solution to load before rather than separate projects one-by-one.

jainaashish replied:
I think we even don't need dte project for this command, it just write out project name, type, and full path which are easily accessible in deferred mode. 

You may not need to fix it right away, but may be, add a TODO to do it later.

alpaix replied:
Current contract is to return entire DTE project:
```cs
[OutputType(typeof(EnvDTE.Project))]
```
I don't think we should break it.

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs:
```diff
@@ -342,16 +342,20 @@ public sealed class UIActionEngine
+                .Projects
+                .Where(project =>
+                    project.ProjectStyle == ProjectModel.ProjectStyle.PackagesConfig &&
+                    project.ProjectServices.Capabilities.SupportsPackageReferences);
```
this might not be sufficient since by default ProjectStyle in MSBuildNuGetProject is set to unknown. So make sure all packages.config projects have this correct style set or continue to check the type cast

alpaix replied:
`MSBuildNuGetProject` initializes its `ProjectStyle` with `ProjectStyle.PackagesConfig` unconditionally. So checking the style here is equivalent to checking the type cast which is less efficient.

jainaashish replied:
this is what i saw in MSBuildNuGetProject currently -
 `metadata.ProjectStyle = ProjectStyle.Unknown;`

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Handlers/ProjectRetargetingHandler.cs:
```diff
@@ -192,7 +192,7 @@ int IVsTrackProjectRetargetingEvents.OnRetargetingAfterChange(string projRef, IV
             });
 #if !VS14
-            if (retargetedProject != null && retargetedProject is LegacyCSProjPackageReferenceProject)
+            if (retargetedProject != null && retargetedProject is LegacyPackageReferenceProject)
```
The `null` check is superfluous here and [below](https://github.com/NuGet/NuGet.Client/pull/1379/files#diff-945f7b5439b881dd38bbed2462bfc761R300).  The `is` operator will return `false` if the left-hand side of the operator is `null`.

https://docs.microsoft.com/en-us/dotnet/articles/csharp/language-reference/keywords/is#testing-for-type-compatibility

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -383,68 +352,68 @@ public bool IsSolutionAvailable
 
+#if VS14
         public async Task<bool> SolutionHasDeferredProjectsAsync()
         {
```
why change this? and have duplicate method signature for VS14 n others?

alpaix replied:
Looks like a merge issue. Will clean this up.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -96,44 +101,34 @@ public NuGetProject DefaultNuGetProject
+            Assumes.Present(projectSystemFactory);
+            Assumes.Present(credentialServiceProvider);
+            Assumes.Present(logger);
+            Assumes.Present(settings);
```
Add a `null` check for `_vsProjectAdapterProvider`.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -832,24 +850,24 @@ private void AddEnvDTEProjectToCache(EnvDTE.Project envDTEProject)
 
-            ProjectNames envDTEProjectName;
-            _projectSystemCache.TryGetProjectNames(name, out envDTEProjectName);
+            ProjectNames projectName;
```
`projectNames`

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProjectProvider.cs:
```diff
@@ -0,0 +1,171 @@
+                Assumes.Present(projectServices);
+
+                return new DeferredProjectServicesProxy(
+                    vsProjectAdapter,
```
would we need to convert it to real VSProjectSystemService? are we already doing it when project gets loaded?

alpaix replied:
When a project gets loaded for real we will create new instance of `NuGetProject` in response to the project-added event.

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProjectProvider.cs:
```diff
@@ -0,0 +1,171 @@
+namespace NuGet.PackageManagement.VisualStudio
+{
+    [Export, Export(typeof(INuGetProjectProvider))]
+    [Name(nameof(LegacyPackageReferenceProjectProvider))]
```
why 2 export?

alpaix replied:
not needed

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProjectProvider.cs:
```diff
@@ -0,0 +1,171 @@
+                }
+
+                var projectServices = componentModel.GetService<VsProjectSystemServices>();
+                Assumes.Present(projectServices);
```
what is the use of this code?

alpaix replied:
not needed.

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapterProvider.cs:
```diff
@@ -0,0 +1,169 @@
+    internal class VsProjectAdapterProvider : IVsProjectAdapterProvider
+    {
+        private readonly IDeferredProjectWorkspaceService _workspaceService;
+        private readonly IVsProjectThreadingService _threadingService;
```
lazy import? since DeferredProjectWorkspaceService should only be initialized when solution is deferred load.

alpaix replied:
Good idea!

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/NuGetProjectTelemetryService.cs:
```diff
@@ -64,21 +64,21 @@ public void EmitNuGetProject(NuGetProject nuGetProject)
                 var projectType = NuGetProjectType.Unknown;
-                if (nuGetProject is MSBuildNuGetProject)
+                if (nuGetProject.ProjectStyle == ProjectModel.ProjectStyle.PackagesConfig)
                 {
```
why not keep consistency either use type cast or project style? And if you use projectstyle then make sure it's being set correctly for all the projects.

alpaix replied:
This PR doesn't claim to make everything consistent. These changes are made to avoid loading deferred projects where it's possible.

jainaashish replied:
I understand, but i find it weird as in same API, we use different mechanisms to check nuget project type, which is little more confusing. Why don't we just use ProjectStyle to check all project types?

alpaix replied:
You're right. I'm about to revert it.

alpaix committed:
Cumulative PR feedback plus bug fixes

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -190,48 +185,31 @@ public NuGetProject DefaultNuGetProject
-            var nuGetProject = await NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async () =>
-            {
-                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+            _projectSystemCache.TryGetProjectNames(projectName, out ProjectNames oldProjectName);
```
`projectNames` and `oldProjectNames` --- they're both of type `ProjectNames`.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -383,68 +352,68 @@ public bool IsSolutionAvailable
-        public async Task<IEnumerable<string>> GetDeferredProjectsFilePathAsync()
-        {
 #if VS14
+        public Task<IEnumerable<string>> GetDeferredProjectsFilePathAsync()
```
Can simplify this by merging with the method below.  Inside the method below put:

```
#if VS14
    // Not applicable for Dev14 so always return empty list.
    return Enumerable.Empty<string>();
#endif
```

Same feedback applies to other methods below.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -664,14 +661,14 @@ private void OnEnvDTEProjectRenamed(EnvDTE.Project envDTEProject, string oldName
+                    RemoveVsProjectAdapterFromCache(oldName);
+                    AddVsProjectAdapterToCache(_vsProjectAdapterProvider.CreateVsProject(envDTEProject));
                     NuGetProject nuGetProject;
                     _projectSystemCache.TryGetNuGetProject(envDTEProject.Name, out nuGetProject);
```
Shouldn't we be checking the return value of `TryGetNuGetProject(...)` before the `out` variable is used below?  The same feedback applies [here](https://github.com/NuGet/NuGet.Client/pull/1379/files#diff-829b08c28853b1bf9bf136e579c869d1R749).

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -745,38 +742,60 @@ private void SetDefaultProjectName()
             {
                 try
                 {
+                    var defferedProjects = GetDeferredProjects();
```
`deferredProjects`

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Migrate/ProjectJsonToPackageRefMigrator.cs:
```diff
@@ -73,23 +79,7 @@ public static class ProjectJsonToPackageRefMigrator
-                }
-
-                await project.InstallPackageWithMetadataAsync(dependency.Name, dependency.LibraryRange.VersionRange, metadataElements, metadataValues);
+                await project.ProjectServices.References.AddOrUpdatePackageReferenceAsync(dependency, CancellationToken.None);
```
Should we plumb through a `CancellationToken` to avoid passing `CancellationToken.None` here?

alpaix replied:
Good point! I planned to limit scope of the change here. I'll consider propagating `CancellationToken` up to the entry point.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/DeferredProjectServicesProxy.cs:
```diff
@@ -0,0 +1,86 @@
+        {
+            get
+            {
+               return _threadingService.JoinableTaskFactory.Run(async delegate
```
Does a project's `IsDeferred` property return a value which may change over time?  If no, consider making this lazy to memoize the result.

alpaix replied:
It's a dynamic property so we'd rather not cache it.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemService.cs:
```diff
@@ -0,0 +1,218 @@
+            Assumes.Present(projectServices);
+
+            _vsProjectAdapter = vsProjectAdapter;
+            Assumes.Present(_vsProjectAdapter);
```
Superfluous because you already checked `vsProjectAdapter` above.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/DeferredProjectWorkspaceService.cs:
```diff
@@ -59,51 +60,29 @@ internal sealed class DeferredProjectWorkspaceService : IDeferredProjectWorkspac
+                        { "TargetFramework", targetFramework }
+                    };
+                return await factory.GetMSBuildProjectDataServiceAsync(
+                    projectFilePath, projectProperties: projectProperties);
```
The `projectProperties` named parameter is superfluous because the [local variable](https://github.com/NuGet/NuGet.Client/pull/1379/files#diff-6d9d2e798141f06ef9e080f10e94a989R79) is aptly named.

alpaix replied:
The factory method has more parameters with default values. This statement selectively overrides one of them.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -932,25 +947,25 @@ private NuGetProject CreateNuGetProject(Project envDTEProject, INuGetProjectCont
-        internal static IEnumerable<EnvDTE.Project> GetDependentEnvDTEProjects(IDictionary<string, List<EnvDTE.Project>> dependentEnvDTEProjectsDictionary, EnvDTE.Project envDTEProject)
+        internal static IEnumerable<IVsProjectAdapter> GetDependentProjects(IDictionary<string, List<IVsProjectAdapter>> dependentProjectsDictionary, IVsProjectAdapter vsProjectAdapter)
         {
             Debug.Assert(ThreadHelper.CheckAccess());
```
Should this apply to retail builds as well?

    ThreadHelper.ThrowIfNotOnUIThread();

You made a similar change [here](https://github.com/NuGet/NuGet.Client/pull/1379/files#diff-117f4f7f52c93efcd353cce4f5f9e13eR46).

alpaix replied:
Correct, it's hard to change every occurrence. I might've missed hundreds of them :(

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -932,25 +947,25 @@ private NuGetProject CreateNuGetProject(Project envDTEProject, INuGetProjectCont
         }
 
         // REVIEW: This might be inefficient, see what we can do with caching projects until references change
```
Does this comment still apply?  The current method implementation looks reasonably performant in that it's already getting dependent projects from a dictionary cache.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectBuildProperties.cs:
```diff
@@ -0,0 +1,52 @@
+            EnvDTE.Project project,
+            IVsProjectThreadingService threadingService)
+        {
+            Assumes.Present(project);
```
`null` check `threadingService`.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsLangProjectBuildProperties.cs:
```diff
@@ -0,0 +1,57 @@
+            _threadingService = threadingService;
+        }
+
+        public string GetPropertyValue(string propertyName)
```
Can `propertyName` be `null` or an empty string?

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsLangProjectSystemService.cs:
```diff
@@ -0,0 +1,287 @@
+            IVsProjectAdapter projectAdapter,
+            INuGetProjectServices projectServices)
+        {
+            Assumes.Present(projectAdapter);
```
`null` check `projectServices`.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsLangProjectSystemService.cs:
```diff
@@ -0,0 +1,287 @@
+            _threadingService = projectServices.GetGlobalService<IVsProjectThreadingService>();
+            Assumes.Present(_threadingService);
+
+            _asVSProject4 = new Lazy<VSProject4>(() => _project.Object as VSProject4);
```
Why is this lazy?  I'm sure you have a good reason; I'd just like to understand it.

alpaix replied:
We use Lazy freely wherever it's possible. The reason here is to defer accessing shell object until it's really needed and when switching to the main thread will be possible (not in the ctor).

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsLangProjectSystemService.cs:
```diff
@@ -0,0 +1,287 @@
+        }
+
+        public async Task<IEnumerable<LibraryDependency>> GetPackageReferencesAsync(
+            NuGetFramework targetFramework, CancellationToken _)
```
Check arguments?

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsLangProjectSystemService.cs:
```diff
@@ -0,0 +1,287 @@
+                    Array metadataValues;
+                    AsVSProject4.PackageReferences.TryGetReference(installedPackage, _referenceMetadata, out version, out metadataElements, out metadataValues);
+
+                    return new PackageReference(
```
What does it mean to return a `PackageReference` if the preceding call to `TryGetReference(...)` fails?  You'd have a `PackageReference` with `null` `version`, `metadataElements`, and `metadataValues`.  Is that reasonable?

alpaix replied:
Good point.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsLangProjectSystemService.cs:
```diff
@@ -0,0 +1,287 @@
+            return string.Empty;
+        }
+
+        public async Task AddOrUpdatePackageReferenceAsync(LibraryDependency packageReference, CancellationToken _)
```
`null` check `packageReference`.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsLangProjectSystemService.cs:
```diff
@@ -0,0 +1,287 @@
+                metadataValues);
+        }
+
+        public void RemovePackageReference(string packageName)
```
Check argument for `null` or empty string.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsLangProjectSystemService.cs:
```diff
@@ -0,0 +1,287 @@
+                TargetNuGetFramework = targetNuGetFramework;
+            }
+
+            public string Name { get; private set; }
```
Consider removing the private setters to enforce readonly semantics.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsProjectScriptHostService.cs:
```diff
@@ -0,0 +1,103 @@
+            var scriptExecutor = _scriptExecutor.Value;
+            Assumes.Present(scriptExecutor);
+
+            var packageReader = new PackageFolderReader(packageInstallPath);
```
Wrap `packageReader` in a `using` statement.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/WorkspaceProjectServices.cs:
```diff
@@ -0,0 +1,130 @@
+        {
+            if (item.Metadata.Keys.Contains(propertyName))
+            {
+                return item.Metadata[propertyName];
```
Can you simplify 2 lookups into a single `TryGetValue(...)` call?

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:
```diff
@@ -270,7 +246,7 @@ public void AddExistingFile(string path)
             if (!File.Exists(fullPath))
             {
-                throw new ArgumentNullException(String.Format(CultureInfo.CurrentCulture, Strings.PathToExistingFileNotPresent, fullPath, ProjectName));
+                throw new ArgumentNullException(string.Format(CultureInfo.CurrentCulture, Strings.PathToExistingFileNotPresent, fullPath, ProjectName));
```
The `ArgumentNullException(string)` constructor takes the parameter name not a message.  What should be thrown here is a `FileNotFoundException`.

alpaix replied:
Seems odd, no doubt. Although I'm not sure what protocol expectations are made by its callers. So I'd prefer not touching this.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:
```diff
@@ -564,33 +376,9 @@ public virtual void RemoveFile(string path)
         public virtual void RemoveImport(string targetFullPath)
         {
-            if (String.IsNullOrEmpty(targetFullPath))
+            if (string.IsNullOrEmpty(targetFullPath))
```
`Assumes.NotNullOrEmpty(targetFullPath);`?

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:
```diff
@@ -990,17 +711,267 @@ public IEnumerable<string> GetFullPaths(string fileName)
+        public Task<IEnumerable<ProjectRestoreReference>> GetProjectReferencesAsync(
+            Common.ILogger logger, CancellationToken _)
+        {
+            return Task.FromResult<IEnumerable<ProjectRestoreReference>>(Enumerable.Empty<ProjectRestoreReference>().ToList());
```
Is it necessary to call `ToList()`?

alpaix replied:
Not anymore

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:
```diff
@@ -990,17 +711,267 @@ public IEnumerable<string> GetFullPaths(string fileName)
+
+            NuGetProjectContext.Log(
+                ProjectManagement.MessageLevel.Debug,
+                $"Added reference '{name}' to project:'{projectName}'. Was the Reference Resolved To Package (resolvedToPackage):'{resolvedToPackage}', " +
```
Should this be a string resource?

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:
```diff
@@ -990,17 +711,267 @@ public IEnumerable<string> GetFullPaths(string fileName)
+
+                return References.Item(referenceName) != null;
+            }
+            catch
```
Why not log this exception?

alpaix replied:
I think there were a good reason no to log (too noisy?). It's a reference query that should be swallowing the exception.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/INuGetProjectProvider.cs:
```diff
@@ -2,22 +2,33 @@
         /// <summary>
+        /// Type of project this provider creates
+        /// </summary>
+        System.Type ProjectType { get; }
```
Consider `System.RuntimeTypeHandle` instead of `System.Type`.

https://blogs.msdn.microsoft.com/vancem/2006/10/01/drilling-into-net-runtime-microbenchmarks-typeof-optimizations/

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:
```diff
@@ -0,0 +1,280 @@
+
+            var projectReferences = await ProjectServices
+                .ReferencesReader
+                .GetProjectReferencesAsync(Common.NullLogger.Instance, CancellationToken.None);
```
This method gets called by `GetPackageSpecsAsync(...)`, which gets called by `DependencyGraphRestoreUtility` methods.  Why not plumb down the `CancellationToken` from there?

`GetPackageSpecsAsync(...)` looks like a large uninterruptible chunk of work.

alpaix replied:
It'd be nice but this API is a cornerstone of V3 restore. And that would greatly enlarge scope of the PR.

dtivel commented on src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs:
```diff
@@ -35,7 +35,7 @@ public class MSBuildProjectSystem : MSBuildUser, IMSBuildNuGetProjectSystem
         }
 
-        public INuGetProjectContext NuGetProjectContext { get; private set; }
+        public INuGetProjectContext NuGetProjectContext { get; set; }
```
I'm not crazy about public read/write properties.  Read-only properties are superior.  Are callers of this property (both internal and external) properly guarding against this property being null?  At least with the old `SetNuGetProjectContext(...)` method, the method could disallow `null` values.

alpaix replied:
This makes sense. Although existing code didn't enforce null-check for this property. I'd avoid changing this behavior in the scope of this PR.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsProjectSystemServices.cs:
```diff
@@ -0,0 +1,80 @@
+            }
+        }
+
+        public IProjectSystemCapabilities Capabilities { get; set; }
```
My [previous comment](https://github.com/NuGet/NuGet.Client/pull/1379/files#diff-dd57943aa0889dab1133c1101a8c6b1aR38) applies here as well.  Can the setters of these properties be `private`?

For example, [this location](https://github.com/NuGet/NuGet.Client/pull/1379/files#diff-90c1fab45ee433893a3a539f7fb70882R105) dereferences the `ProjectSystem` property which, from my very cursory examination, could be `null`.  The `ProjectSystem` property is set [here](https://github.com/NuGet/NuGet.Client/pull/1379/files#diff-2e2d8e0a8d42d615042893d1b12fb626R102) but not [here](https://github.com/NuGet/NuGet.Client/pull/1379/files#diff-bd78bca563ac7d501229d97c74bd58c1R133) or [here](https://github.com/NuGet/NuGet.Client/pull/1379/files#diff-12613cf5d824e815347c07c7643618acR164).

You could introduce a method to set these properties and disallow `null` values where it makes sense.  If a property is truly nullable, then it's not clear to me that all call sites are properly checking for `null` values first.

alpaix replied:
Will consider improving the design.

dtivel commented on test/TestExtensions/API.Test/VSSolutionHelper.cs:
```diff
@@ -23,6 +24,13 @@ public static class VSSolutionHelper
 
+        public static void WaitForSolutionLoad()
+        {
+            var mre = new ManualResetEvent(false);
```
`using`

dtivel commented on src/NuGet.Core/NuGet.PackageManagement/Projects/IProjectBuildProperties.cs:
```diff
@@ -0,0 +1,28 @@
+        string GetPropertyValue(string propertyName);
+
+        /// <summary>
+        /// Asyncronous method to retrieve a property value.
```
Asynchronous

dtivel commented on src/NuGet.Core/NuGet.PackageManagement/Projects/DefaultProjectServices.cs:
```diff
@@ -0,0 +1,109 @@
+
+        public string GetPropertyValue(string propertyName)
+        {
+            return null;
```
It's not clear what the pattern is:  do methods return a default value or throw?  It seems like a mixture of the two.

alpaix replied:
It is a mixture. Its behavior matches default project system implementation before the change.

dtivel commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/PreinstalledPackageInstaller.cs:
```diff
@@ -47,13 +48,15 @@ internal class PreinstalledPackageInstaller
         {
             //_websiteHandler = websiteHandler;
             _packageServices = packageServices;
             //_vsCommonOperations = vsCommonOperations;
```
Remove comments?

dtivel commented on src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectSystems/IProjectSystemCache.cs:
```diff
@@ -101,7 +101,7 @@ public interface IProjectSystemCache
         /// <param name="dteProject">The VS project.</param>
```
Update

dtivel commented on src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectSystems/IProjectSystemCache.cs:
```diff
@@ -42,7 +42,7 @@ public interface IProjectSystemCache
         /// <param name="name">Project name, full path or unique name.</param>
         /// <param name="dteProject">Desired project object, not initialized if not found.</param>
```
Update

dtivel reviewed (approved):
I'd like to see some end-to-end test coverage added for LSL support, though it need not be part of this PR.

jainaashish reviewed (approved):
Few minor comments, but overall looks good. Also, it will be good if we can add some functional or end to end tests

alpaix commented:
Issue tracking end-to-end test work NuGet/Home#5258.

alpaix committed:
Applied recommended design changes

alpaix committed:
Merge branch 'dev' into dev-lslIntergated

alpaix committed:
[BUG 441140] An error dialog pop up when creating C# ASP.NET Sigle Page Application

Fixes internal bug 441140.

alpaix committed:
Merge branch 'dev' into dev-lslIntergated
